core/src/main/scala/org/bitcoins/core/hd/HDPath.scala

```diff
@@ -0,0 +1,36 @@
+package org.bitcoins.core.hd
+import scala.util.Try
+
+private[bitcoins] trait HDPath[T <: HDPath[_]] extends BIP32Path {
+
+  /**
+    * Increments the address index and returns the
+    * new path that can be passed into a
+    * [[org.bitcoins.core.crypto.ExtKey ExtKey]]
+    */
+  // TODO check out this cast
```

Comment:
ping

Reply:
This is not ideal. We cannot match on `T`. In the same vein we cannot introduce a type param to `toPath` and match on that either. However in `toPath` we match on the `purpose` field (which is the level that defines the type of path we have, legacy/segwit/nested-segwit). It's impossible to construct a segwit path with a non-segwit purpose field, and similarly with the other types of paths. Therefore I _think_ this cast is always safe. Are you fine with leaving it in, if I also add a comment something along the line of what I'm writing here? 

I'm not really sure how to handle it without the cast.